### ISSUE TITLE
Continuous Pull Replication Moves to 'Idle' Before All Documents Pulled Down

### ISSUE THREAD

**Issue by tikimcfee** [couchbase/couchbase-lite-android#655]
Hey there guys,

I'm working on revamping our app's synchronization flow and its handling of pull / push replicators. Specifically, I'm working on our case where we want to do a large initial pull of all of a user's documents, and upon completion, move to the next phase of app loading.

The issue seems to be with continuous replications and how their status changes come through on their change listeners. Per documentation, I've read from [Observing and monitoring replications](http://developer.couchbase.com/documentation/mobile/current/develop/guides/couchbase-lite/native-api/replication/index.html):

> Idle: Indicates that a continuous replication has "caught up" and transferred all documents, but is
> monitoring the source database for future changes.

What I've found is the the `Idle` status comes back intermittently throughout the pull replication, and is not an accurate accounting of when the replicator is *actually* finished. The current workaround in place is to use `setContinuous(false)` to get a `oneShot` replication, and then switching the replicator to continuous when it (correctly) comes back with a `Stopped` status. This has the issue of the replicator doing a full handshake pass on each of its documents that it just pulled to ensure there are no new revisions. The iOS team I work with as well has shown that this is generally *not* the case with the continuous replicators on their platform, and the `Idle` state does come back when the pull replicator has caught up with our gateway.

This takes a long time to do for some of our users' accounts, and just generally makes for a degraded experience when using the app. It would also end up making our development and error handling flow a *lot* cleaner to know that we can rely on just one set of replicators / statuses!

Is this a known issue / process of pull replication handling? Or, is there something else we should be keying off of to determine that a continuous pull replicator is buttoned-up-and-really-truly-idle? Some digging around in the libs show a sorta-documented by specifically *not* public API for something called ReplicationStateTransition and ReplicationTrigger. Does our answer lie in there somewhere?

**Comment by hideki** [comment]
Hi @tikimcfee,

Thank you very much for you bug report.

> Is this a known issue / process of pull replication handling? 

We have been working to fix this issue. Actually we could solve this issue for some user scenarios. But, not for all cases. I believe this issue depends on number of documents, size of document, and device type, and so on. If you could provide us specific condition, we will try to identify the problem.

Thanks,
Hideki

**Comment by tikimcfee** [comment]
Thanks for the uber-quick response on this @hideki !

I just heard from my teammate Justin Camp that you actually worked with us on our project on a few other issues. If you would, drop me a line at iheartlappy486@gmail.com and I can help hook you up with a repro path for this.

**Comment by zgramana** [comment]
@tikimcfee we'd love to hear back from you.

**Comment by hideki** [comment]
We have access to test sync gateway now. We should tackle this issue during next sprint.

**Comment by hideki** [comment]
This issue is duplicated with https://github.com/couchbase/couchbase-lite-java-core/issues/1077. Issue will be tracked by other ticket. So I will close this.

**Comment by rajasaur** [comment]
Hi @tikimcfee , We have a similar issue (and for a long time were just looking at our code to see if it was our fault) but just saw this. I tried the approach you did (OneShot first and waiting for Stopped), but even then I dont get the whole set of documents.

We have about 90000+ documents and we use a replication filter when pulling documents from couchdb, it so happens that after the first 3500 documents, it just goes into Stopped state and we get the same problem as before, just a bit more delayed (When we do continuous replication, it used to stop at about 2300 documents and say that its done). Is there something else that you did to get over this problem while the main issue is fixed in the core library?

Thanks!

**Comment by tikimcfee** [comment]
Hey @rajasaur 
Sorry for the late response to ya - I'm glad you found a solution!

To answer your original question, ` Is there something else that you did to get over this problem...` we did eventually add a 'guard' mechanism to our app launch process that tries to make sure a core set of documents are available _on disk_ before moving on to a main activity.

It's actually pretty simple : Attach a listener to your instance of the CBLite database. Test load some documents that are interdependent. If a doc doesn't comes back with its full set of properties, wait for the listener to tell you when it has changed, and see if any of those changes fulfill your missing dependencies, and repeat. This was the path we took when we had a core set of documents that _had_ to be available for a new feature to function correctly.
